### PR TITLE
use different lifetime name for object-lifetime-default elision

### LINKED ISSUE #62517: async fn with elided lifetime causes rustc panic
Code to reproduce:

```rust
use futures::Stream;
use std::any::Any;

async fn foo(x: &str) -> impl Stream<Item = Box<dyn Any>> {}
```

and then run `cargo check`... (with full backtrace)

```
thread 'rustc' panicked at 'assertion failed: !erased_self_ty.has_escaping_bound_vars()', src/librustc/ty/util.rs:356:9
stack backtrace:
   0:        0x1054341e2 - std::panicking::default_hook::{{closure}}::h732feb85026a4422
   1:        0x105433eab - std::panicking::default_hook::h832638be7920e1d9
   2:        0x10410d3a3 - rustc::util::common::panic_hook::h50956d861d3d8265
   3:        0x105434aa1 - std::panicking::rust_panic_with_hook::h2746b02166e15af2
   4:        0x10409f2a5 - std::panicking::begin_panic::h230090070c2be599
   5:        0x104015c5d - rustc::ty::util::<impl rustc::ty::context::TyCtxt>::required_region_bounds::h2bfd0992e0240e65
   6:        0x1041b295b - rustc::infer::opaque_types::Instantiator::fold_opaque_ty::h6a2109340442464e
   7:        0x1042616a2 - <rustc::ty::fold::BottomUpFolder<F,G,H> as rustc::ty::fold::TypeFolder>::fold_ty::h7aa13d9636d94705
   8:        0x103e0765a - rustc::ty::fold::TypeFoldable::fold_with::h6eb10112104db7f6
   9:        0x1041b2d17 - rustc::infer::opaque_types::Instantiator::fold_opaque_ty::h6a2109340442464e
  10:        0x102eee399 - <rustc::ty::fold::BottomUpFolder<F,G,H> as rustc::ty::fold::TypeFolder>::fold_ty::h362061237fa87a5f
  11:        0x102e9619f - rustc_typeck::check::FnCtxt::instantiate_opaque_types_from_value::hdd338307918cfee6
  12:        0x102e8bcaa - rustc_typeck::check::check_fn::hac7bd15dbe73baf3
  13:        0x102fd07f4 - rustc::ty::context::GlobalCtxt::enter_local::h324ac2ec2e12e0e2
  14:        0x102e8b1ac - rustc_typeck::check::typeck_tables_of::h8a6866c5ac4e4538
  15:        0x102eab71e - rustc::ty::query::__query_compute::typeck_tables_of::h461c35490ac0925d
  16:        0x102f4d49b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute::hb6ba7a7a89baca84
  17:        0x102edc256 - rustc::dep_graph::graph::DepGraph::with_task_impl::h729ba93c71ed6836
  18:        0x102facd48 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::ha9abd6cc76593f28
  19:        0x102eab71e - rustc::ty::query::__query_compute::typeck_tables_of::h461c35490ac0925d
  20:        0x102f4d49b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::typeck_tables_of>::compute::hb6ba7a7a89baca84
  21:        0x102edc256 - rustc::dep_graph::graph::DepGraph::with_task_impl::h729ba93c71ed6836
  22:        0x102facd48 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::ha9abd6cc76593f28
  23:        0x103017d8c - rustc_typeck::collect::checked_type_of::h6778c833e01799c6
  24:        0x10301781e - rustc_typeck::collect::type_of::h1264787cfb7a8feb
  25:        0x102ead1cb - rustc::ty::query::__query_compute::type_of::h8914623b8a89443b
  26:        0x102f4d0cb - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::type_of>::compute::h9e129d1cba94e459
  27:        0x102ed5816 - rustc::dep_graph::graph::DepGraph::with_task_impl::h2b7ac820718424f6
  28:        0x102f86f98 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::h5ff957fa238b670a
  29:        0x102ff2a47 - rustc::hir::intravisit::walk_expr::hb0db1b8ea3eea671
  30:        0x103005314 - rustc::hir::intravisit::Visitor::visit_fn::hc5e7e3e1abe9958f
  31:        0x102fe36c3 - rustc::hir::intravisit::walk_impl_item::h4d45bdab776530c9
  32:        0x102eae0c8 - rustc::hir::map::Map::visit_item_likes_in_module::ha9769faef02af1e0
  33:        0x103013444 - rustc_typeck::collect::collect_mod_item_types::h2ec1a0c87fa3d49d
  34:        0x102eac5de - rustc::ty::query::__query_compute::collect_mod_item_types::heaa821ee7f695a15
  35:        0x102f4d75b - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::collect_mod_item_types>::compute::h5f889db1173afe74
  36:        0x102ed5e38 - rustc::dep_graph::graph::DepGraph::with_task_impl::h2ef3fa01ca8acb95
  37:        0x102f7346e - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::h308b4698a27c8896
  38:        0x102ecb422 - rustc_typeck::check_crate::{{closure}}::{{closure}}::h9655a64665552d1e
  39:        0x102eae94d - rustc::util::common::time::h272405b2b749ed07
  40:        0x103058fee - rustc_typeck::check_crate::h484f134202177825
  41:        0x102acdfdf - rustc_interface::passes::analysis::hd60757f4c2b698a8
  42:        0x102a5f012 - rustc::ty::query::__query_compute::analysis::h78ec8ee67474784a
  43:        0x102a60fe5 - rustc::dep_graph::graph::DepGraph::with_task_impl::h4ac160ddc37c9308
  44:        0x1029de862 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query::he04d60713b4447ef
  45:        0x1029f9bec - rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}::h056779aae72902db
  46:        0x102b10acb - rustc_interface::passes::create_global_ctxt::{{closure}}::hfa8388d59c8efa62
  47:        0x1029fc489 - rustc_interface::interface::run_compiler_in_existing_thread_pool::hff72514f7ca933dc
  48:        0x102a0c9a4 - std::thread::local::LocalKey<T>::with::h91a2674509ca585e
  49:        0x102a1cbb2 - scoped_tls::ScopedKey<T>::set::h03f39cddb2457ca3
  50:        0x102a34245 - syntax::with_globals::h6d4146eff33e9bc1
  51:        0x102a569ba - std::sys_common::backtrace::__rust_begin_short_backtrace::hc2071b1843657bff
  52:        0x105443f3f - __rust_maybe_catch_panic
  53:        0x1029e4237 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h1ae87809d2a636db
  54:        0x105417cce - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hffc2c905ed2b6268
  55:        0x105442d5e - std::sys::unix::thread::Thread::new::thread_start::h6ca51316935dd8c4
  56:     0x7fff5b16f2eb - _pthread_body
  57:     0x7fff5b172249 - _pthread_start
query stack during panic:
#0 [typeck_tables_of] processing `s3::Foo::bar`
#1 [typeck_tables_of] processing `s3::Foo::bar::{{closure}}#0`
#2 [type_of] processing `s3::Foo::bar::{{closure}}#0`
#3 [collect_mod_item_types] collecting item types in module `s3`
#4 [analysis] running analysis passes on this crate
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.38.0-nightly (78ca1bda3 2019-07-08) running on x86_64-apple-darwin

note: compiler flags: -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden
```

This also reproduces when `&str` is replaced with `&_` (where `_` is any type), but does't reproduce with `&'static _`.

### PR DESCRIPTION
**Pull request by nikomatsakis**
Introduce a distinct value for `LifetimeName` to use when this is a object-lifetime-default elision. This allows us to avoid creating incorrect lifetime parameters for the opaque types that result. We really need to overhaul this setup at some point! It's getting increasingly byzantine. But this seems like a relatively... surgical fix.

r? @cramertj 

Fixes #62517

### CODE REVIEW

File: src/test/ui/async-await/issue-62517.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// build-pass
+
+#![feature(async_await)]
+
+trait FirstTrait {}
+trait SecondTrait {
+    type Item: ?Sized;
+}
+
+async fn foo(x: &str) -> impl SecondTrait<Item = dyn FirstTrait> {
```
Comment (Centril) [inline src/test/ui/async-await/issue-62517.rs]: Can you also include the version with `-> impl Alpha<dyn Object>` for completeness (so we also cover type parameters and not just associated types)?

File: src/test/ui/async-await/issue-62517.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// Regression test for #62517. We used to ICE when you had an `async
+// fn` with an `impl Trait` return that mentioned a `dyn Bar` with no
+// explicit lifetime bound.
+//
+// edition:2018
+// build-pass
```
Comment (Centril) [inline src/test/ui/async-await/issue-62517.rs]:
```suggestion
// check-pass
```

File: src/test/ui/impl-trait/dyn-trait-elided-two-inputs.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Test that we don't get an error with `dyn Bar` in an impl Trait
+// when there are multiple inputs.  The `dyn Bar` should default to `+
+// 'static`. This used to erroneously generate an error (cc #62517).
+//
+// build-pass
```
Comment (Centril) [inline src/test/ui/impl-trait/dyn-trait-elided-two-inputs.rs]:
```suggestion
// check-pass
```

File: src/librustc/hir/lowering.rs [hunk 4]
```diff
@@ -1908,6 +1933,11 @@ impl<'a> LoweringContext<'a> {
                         }
                     }
                     hir::LifetimeName::Param(_) => lifetime.name,
+
+                    // refers to some other lifetime that is "in scope" within the
```
Comment (Centril) [inline src/librustc/hir/lowering.rs]:
```suggestion
                    // Refers to some other lifetime that is "in scope" within the
```

File: src/librustc/hir/mod.rs [hunk 5]
```diff
@@ -222,6 +222,19 @@ pub enum LifetimeName {
     /// User wrote nothing (e.g., the lifetime in `&u32`).
     Implicit,
 
+    /// Implicit lifetime in a context like `dyn Foo`. This is
+    /// distinguished from implicit lifetimes elsewhere because the
+    /// lifetime that they default to must appear elsewhere within the
+    /// enclosing type.  This means that, in an impl trait context, we
```
Comment (Centril) [inline src/librustc/hir/mod.rs]:
```suggestion
    /// enclosing type.  This means that, in an `impl Trait` context, we
```

File: src/test/ui/impl-trait/dyn-trait-elided-two-inputs.rs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+trait Foo { type Item: ?Sized; }
+trait Bar { }
+
+impl<T> Foo for T {
+    type Item = dyn Bar;
+}
+
+fn foo(x: &str, y: &str) -> impl Foo<Item = dyn Bar> { () }
```
Comment (Centril) [inline src/test/ui/impl-trait/dyn-trait-elided-two-inputs.rs]: Can you also include the type parameter version (`Foo<Item>`) here also?

Comment (Centril) [reply src/test/ui/impl-trait/dyn-trait-elided-two-inputs.rs]: And as a suggestion to apply... :)

```suggestion
fn foo(x: &str, y: &str) -> impl Foo<Item = dyn Bar> { () }

trait Alpha<Item: ?Sized> {}
trait Object {}

impl<T> Foo<dyn Object> for T {}

fn alpha(x: &str, y: &str) -> impl Alpha<dyn Object> { () }

```

File: src/test/ui/async-await/issue-62517.rs (hunk 1 above)
Comment (Centril) [reply src/test/ui/async-await/issue-62517.rs]: Suggestion to apply in https://github.com/rust-lang/rust/pull/63376#discussion_r311926986

**Comment by nikomatsakis** [comment]
Note to self, as I'm not really at keyboard now: 

Check what happens with `&impl Trait<Item = &u32>`, and what we *expect* to happen. =) 

Also check similar scenarios like `&dyn Iterator<Item = &u32>`.

Arguably, we *should* create a lifetime parameter for the opaque type that we create, but it should get a value of `'static` -- i.e., it should follow the "object lifetime default" rules to determine its value. That would be a slightly different fix than what we have here (though probably related).

**Comment by cramertj** [comment]
> Check what happens with &impl Trait<Item = &u32>, and what we expect to happen. =)

Yeah, I'd also be curious about `-> impl Trait<Item = Box<dyn OtherTrait<impl ThirdTrait>>>`.

This fix makes sense to me, but the collection of lifetime-shifting-in-lowering logic is growing increasingly complex and scary. I don't have a particular fix in mind, but spending some time brainstorming could be good.

r=me with those tests added and @Centril's nits addressed.

**Comment by nikomatsakis** [comment]
@cramertj 

> This fix makes sense to me, but the collection of lifetime-shifting-in-lowering logic is growing increasingly complex and scary. I don't have a particular fix in mind, but spending some time brainstorming could be good.

Yes.

> Yeah, I'd also be curious about `-> impl Trait<Item = Box<dyn OtherTrait<impl ThirdTrait>>>`.

I was reviewing our object lifetime default rules (specified in [RFC 599](https://rust-lang.github.io/rfcs/0599-default-object-bound.html) and then amended in [RFC 1156](https://rust-lang.github.io/rfcs/1156-adjust-default-object-bounds.html)). The latter RFC in particular institutes the rule that we take the default from the "innermost enclosing type", essentially; hence `&'x dyn Trait` defaults to `&'x (dyn Trait + 'x)` but `&'x Box<dyn Trait>` defaults to `&'x Box<dyn Trait + 'static>`. This was in response to massive user confusion.

It seems to me that similar logic applies to `&impl Foo<Bar = dyn Trait>` -- that is, we should make the default in this case be the same as it would be for `impl Foo<Bar = dyn Trait>`, and hence it should default to `dyn Trait + 'static`. This is the behavior that the current PR performs, though I think I'd like to review the tests and make sure they are adequate.

Your example:

```rust
-> impl Trait<Item = Box<dyn OtherTrait<impl ThirdTrait>>>
```

would therefore default to:

```rust
-> impl Trait<Item = Box<dyn OtherTrait<impl ThirdTrait> + 'static>>
```

**Comment by nikomatsakis** [comment]
So I added tests for `&impl Foo<Item = dyn Bar>` and, sure enough, encountered problems. The current code for handling object-lifetime defaults basically *overlooking* `Item = XX` bindings, so they de facto just "inherit" the bound from the surrounding context. 

On nightly, this means that `&dyn Foo<Item = dyn Bar>` expands to `&'x dyn Foo<Item = dyn Bar + 'x>` ([example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9b69da08460ddea0f81ef7bbbf73587e)). The behavior for `&impl` is rather inconsistent: on nightly it seems to (de facto) default to `'static`, though I'm not entirely sure why ([example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b7c5025581ad86962117b5d5e01a22ff)). On my branch, however, it would ICE because you wind up with `&'x impl Foo<Item = dyn Bar + 'x>`, and we've not created a lifetime parameter for the `'x` (because of the other changes this branch makes).

I've got a fix that I just pushed but it's not really the right fix. It simply uses `'static` as the default object lifetime bound in `Item = XX` bindings all the time. 

Probably the *right* behavior is to operate analogously an input type parameter. In other words, if you have

```rust
trait Foo<'a, T: 'a + ?Sized> { .. }
```

and `dyn Foo<'x, dyn Bar>`, you will get `dyn Foo<'x, dyn Bar + 'x>`. I could imagine then that

```rust
trait Foo<'a> {
  type Item: 'a + ?Sized;
}
```

and `dyn Foo<'a, Item = dyn Bar>` would be expected to default to `dyn Foo<'a, Item = dyn Bar + 'a>`. But if you have

```rust
trait Foo {
   type Item: ?Sized;
}
```

then `&dyn Foo<Item = dyn Bar>` should default to `dyn Bar + 'static` (unlike today).

**Comment by nikomatsakis** [comment]
Another, conservative option would be to require explicit bounds in this scenario.

**Comment by nikomatsakis** [comment]
I've been investigating what it will take to do the more complex handling of object lifetime default bounds that I described above. It would definitely be complex, because at the phase where we are figuring this sort of thing out, we haven't yet resolved the `Item` to a specific associated type id. That is currently handled by a separate bit of code. On the other hand, I don't see any reason why it'd be impossible to do so -- we know the trait def-id, and that should give us everything we need to do the resolution. But it feels like it's going to be a non-trivial patch to me. 

I suppose a conservative option would be to default to `'static` unless the trait has lifetime parameters -- i.e., if there are no lifetime parameters, then we can't have the case I was thinking about above. I'm tinkering with implementation that for now. This should affect relatively little code -- you'd have to do something like `dyn Foo<'a, Item = dyn Bar>`.

**Comment by nikomatsakis** [comment]
I pushed a commit implementing the conservative rule I described above. It's technically a breaking chance, as I noted, although I believe the current behavior is pretty clearly a bug. I wouldn't *expect* a large impact but it's hard to know.

The affected case is something like `&dyn Foo<Item = dyn Bar>` -- i.e., a `&dyn` type that contains a `Item =` binding with a `dyn` type that is not enclosed in some kind of pointer.

**Comment by nikomatsakis** [comment]
@bors try 

In case we want to try and do a crater run

**Comment by Centril** [comment]
@craterbot run mode=check-only

**Comment by nikomatsakis** [comment]
Dear @rust-lang/lang -- this PR makes a breaking change (bug fix, in my opinion) to the compiler. We are doing a crater run now to see if there are any effects. I wrote up [a summary of the situation in a gist](https://gist.github.com/nikomatsakis/e0ac0276744d10b952cf4ea3d3d8a814), explaining both the background, intended behavior, current behavior, and the effect of this PR.

**Comment by Centril** [comment]
@craterbot p=2

File: src/test/ui/async-await/issues/issue-62517-2.rs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+// edition:2018
+
+#![feature(async_await)]
+
+trait Object {}
+
+trait Alpha<Param> {}
+
+async fn foo<'a>(_: &'a ()) -> impl Alpha<dyn Object> {}
+//~^ ERROR not satisfied
```
Comment (Centril) [inline src/test/ui/async-await/issues/issue-62517-2.rs]: On re-reading the PR... A small nit: Can we add the necessary impl to make this `// check-pass` instead?

**Review by Centril — APPROVED** [review]
This looks great. Thanks also for the write-up.

**Comment by Centril** [comment]
Regression tree (from root):
- [soketto-0.2.2](https://crater-reports.s3.amazonaws.com/pr-63376/try%23266783e4e09e4e9d5307c1c8e695659c58bbcac7/reg/soketto-0.2.2/log.txt)
    - libp2p-websocket-0.11.0
      - libp2p-0.11.0
        - blockchain-network-simple
           ^--- This is the only leaf reverse dependency in the whole tree.

Seems reasonably fixable by just sending a PR.
Imo the single root regression is acceptable enough to land this outright.

File: src/test/ui/async-await/issues/issue-62517-2.rs (hunk 7 above)
Comment (nikomatsakis) [reply src/test/ui/async-await/issues/issue-62517-2.rs]: Done, pushed (at least I think it's done, didn't have time to let tests run to completion yet)

**Pull request merged**